File: app/src/main/java/com/speedchecker/android/sdk/demo/MainActivity.java
package com.speedchecker.android.sdk.demo;

import android.os.Bundle;
import android.text.method.ScrollingMovementMethod;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;

import androidx.appcompat.app.AppCompatActivity;

import com.speedchecker.android.sdk.Public.SpeedTestListener;
import com.speedchecker.android.sdk.Public.SpeedTestResult;
import com.speedchecker.android.sdk.SpeedcheckerSDK;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

public class MainActivity extends AppCompatActivity implements SpeedTestListener {

    private TextView mTextViewStage;
    private TextView mTextViewResult;
    private TextView mTextViewLog;
    private Button mButtonSpeedTest;
    private SimpleDateFormat sdf = new SimpleDateFormat("HH:mm:ss.SSS", Locale.US);

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        mTextViewStage = findViewById(R.id.textView_stage);
        mTextViewResult = findViewById(R.id.textView_result);
        mTextViewLog = findViewById(R.id.textView_log);
        mButtonSpeedTest = findViewById(R.id.button_speedTest);

        mTextViewResult.setMovementMethod(new ScrollingMovementMethod());
        mTextViewLog.setMovementMethod(new ScrollingMovementMethod());

        //TODO: 3. Initialize SpeedcheckerSDK inside a main activity.
        SpeedcheckerSDK.init(this);

        //TODO: 4. Ask Location permissions.
        SpeedcheckerSDK.askPermissions(this);
        //OR request location permissions manually. We need:
        //ACCESS_COARSE_LOCATION
        //ACCESS_FINE_LOCATION
        //ACCESS_BACKGROUND_LOCATION

        //TODO: 6. Set implemented interface.
        SpeedcheckerSDK.SpeedTest.setOnSpeedTestListener(this);

        mButtonSpeedTest.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                //TODO: 7. Start Speed test.
                SpeedcheckerSDK.SpeedTest.startTest(MainActivity.this);
            }
        });

        //===========================
        //=== Additional features ===
        //===========================
        //NOTE: some features could not be available in some SDK versions.
        // Please, contact us to get more information

        //# Interrupt speed test
        //SpeedcheckerSDK.SpeedTest.interruptTest();

        //# Register listener to get additional information (IP, ISP).
        // This callback will fire after internet speed test
        //SpeedcheckerSDK.SpeedTest.setAfterTestUserInfoListener(new SpeedTestListener.AfterTestUserInfo() {
        //    @Override
        //    public void getUserInfoCallback(String IP, String ISP) {
        //
        //    }
        //});
    }

    //TODO: 5. Implement the "SpeedTestListener" interface and override methods.
    @Override
    public void onTestStarted() {
        mTextViewStage.setText("Test Started");
        mTextViewResult.setText("-");
        log("Test Started");
    }

    @Override
    public void onFetchServerFailed() {
        mTextViewStage.setText("Fetch Server Failed");
        mTextViewResult.setText("-");
        log("Fetch Server Failed");
    }

    @Override
    public void onFindingBestServerStarted() {
        mTextViewStage.setText("Finding best server");
        mTextViewResult.setText("...");
        log("Finding best server");
    }

    @Override
    public void onTestFinished(SpeedTestResult speedTestResult) {
        String finalStr =
                "Server: " + speedTestResult.getServer().Domain + "\n"
                        + "Ping: " + speedTestResult.getPing() + " ms" + "\n"
                        + "Download speed: " + speedTestResult.getDownloadSpeed() + " Mb/s" + "\n"
                        + "Upload speed: " + speedTestResult.getUploadSpeed() + " Mb/s" + "\n"
                        + "Connection type: " + speedTestResult.getConnectionTypeHuman() + "\n";
        mTextViewStage.setText("Test Finished");
        mTextViewResult.setText(finalStr);
        log("Test Finished: Server[" + speedTestResult.getServer().Domain + "] -> " + speedTestResult.toString());
    }

    @Override
    public void onPingStarted() {
        mTextViewStage.setText("Ping Started");
        mTextViewResult.setText("...");
        log("Ping Started");
    }

    @Override
    public void onPingFinished(int ping, int jitter) {
        mTextViewStage.setText("Ping Finished");
        mTextViewResult.setText(ping + " ms | jitter: " + jitter);
        log("Ping Finished: " + ping + " ms| jitter: " + jitter);
    }

    @Override
    public void onDownloadTestStarted() {
        mTextViewStage.setText("Download Test Started");
        mTextViewResult.setText("...");
        log("Download Test Started");
    }

    @Override
    public void onDownloadTestProgress(int i, double speedMbs, double transferredMb) {
        mTextViewStage.setText("Download Test Progress");
        mTextViewResult.setText(i + "% -> " + speedMbs + " Mb/s\nTransferredMb: " + transferredMb);
        log("Download Test Progress: " + i + "% -> " + speedMbs + " Mb/s\nTransferredMb: " + transferredMb);
    }

    @Override
    public void onDownloadTestFinished(double v) {
        mTextViewStage.setText("Download Test Finished");
        mTextViewResult.setText(v + " Mb/s");
        log("Download Test Finished: " + v + " Mb/s");
    }

    @Override
    public void onUploadTestStarted() {
        mTextViewStage.setText("Upload Test Started");
        mTextViewResult.setText("...");
        log("Upload Test Started");
    }

    @Override
    public void onUploadTestProgress(int i, double speedMbs, double transferredMb) {
        mTextViewStage.setText("Upload Test Progress");
        mTextViewResult.setText(i + "% -> " + speedMbs + " Mb/s\nTransferredMb: " + transferredMb);
        log("Upload Test Progress: " + i + "% -> " + speedMbs + " Mb/s\nTransferredMb: " + transferredMb);
    }

    @Override
    public void onUploadTestFinished(double v) {
        mTextViewStage.setText("Upload Test Finished");
        mTextViewResult.setText(v + " Mb/s");
        log("Upload Test Finished: " + v + " Mb/s");
    }

    @Override
    public void onTestWarning(String s) {
        mTextViewStage.setText("Test Warning");
        mTextViewResult.setText(s);
        log("Test Warning: " + s);
    }

    @Override
    public void onTestFatalError(String s) {
        mTextViewStage.setText("Test Fatal Error");
        mTextViewResult.setText(s);
        log("Test Fatal Error: " + s);
    }

    @Override
    public void onTestInterrupted(String s) {
        mTextViewStage.setText("Test Interrupted");
        mTextViewResult.setText(s);
        log("Test Interrupted: " + s);
    }

    private void log(String message){
        mTextViewLog.setText("[" + sdf.format(new Date(System.currentTimeMillis())) + "]: "
                + message + "\n" + mTextViewLog.getText());
    }
}

File: app/src/main/java/com/speedchecker/MainActivity.kt
package com.speedchecker

import android.os.Bundle
import android.text.SpannableStringBuilder
import android.text.method.ScrollingMovementMethod
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.core.text.bold
import com.speedchecker.android.sdk.Public.SpeedTestListener
import com.speedchecker.android.sdk.Public.SpeedTestResult
import com.speedchecker.android.sdk.SpeedcheckerSDK
import com.speedchecker.databinding.ActivityMainBinding
import java.text.SimpleDateFormat
import java.util.*

class MainActivity : AppCompatActivity(), SpeedTestListener {

    private lateinit var binding: ActivityMainBinding
    private val sdf = SimpleDateFormat("HH:mm:ss.SSS", Locale.US)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        binding.tvLog.movementMethod = ScrollingMovementMethod()
        binding.testProgressBar.max = 240

        SpeedcheckerSDK.init(this)
        SpeedcheckerSDK.SpeedTest.setOnSpeedTestListener(this)

        binding.btnStart.setOnClickListener {
            SpeedcheckerSDK.SpeedTest.startTest(this@MainActivity)
            binding.btnStart.isEnabled = false
            binding.testProgressBar.progress = 0
            binding.containerResult.visibility = View.INVISIBLE
            binding.containerTest.visibility = View.VISIBLE
        }
    }

    override fun onTestStarted() {
        binding.tvProgress.text = ""
        binding.tvCurrentTestValue.text = ""
        binding.tvTrafficTestValue.text = ""
        binding.ivTestStatusIcon.setImageBitmap(null)
        binding.tvTestStatusText.text = getString(R.string.test_started)
        log(getString(R.string.test_started))
    }

    override fun onFetchServerFailed(errorCode: Int) {
        binding.tvTestStatusText.text = getString(R.string.text_servers_error)
        log(getString(R.string.text_servers_error))
        binding.btnStart.isEnabled = true
    }

    override fun onFindingBestServerStarted() {
        binding.tvTestStatusText.text = getString(R.string.test_fetch_server_progress)
        log(getString(R.string.test_fetch_server_progress))
    }

    override fun onTestFinished(speedTestResult: SpeedTestResult) {
        binding.btnStart.isEnabled = true
        binding.containerTest.visibility = View.INVISIBLE
        binding.containerResult.visibility = View.VISIBLE
        binding.tvServer.text = speedTestResult.server.Domain
        val pingSpan = SpannableStringBuilder()
        pingSpan.bold { append(speedTestResult.ping.toString()) }.append(" ms")
        binding.tvResultPing.text = pingSpan
        binding.tvTestResultDownload.text = "${speedTestResult.downloadSpeed}"
        binding.tvTestResultUpload.text = "${speedTestResult.uploadSpeed}"
        val jitterSpan = SpannableStringBuilder()
        jitterSpan.bold { append(speedTestResult.jitter.toString()) }.append(" ms")
        binding.tvResultJitter.text = jitterSpan
        val indoorStatus =
            if (speedTestResult.indoorOutdoorStatus == 0) getString(R.string.status_outdoor) else getString(
                R.string.status_indoor
            )
        binding.tvLocationType.text = indoorStatus
        speedTestResult.packetLoss?.let {
            binding.tvResultPacketLoss.text = speedTestResult.packetLoss.toString() + "%"
        } ?: run {
            binding.tvResultPacketLoss.text = "-"
        }
        binding.ivTestStatusIcon.setImageBitmap(null)
        binding.tvConnectionType.text = speedTestResult.connectionTypeHuman
        binding.tvTestStatusText.text = getString(R.string.test_finished)
        log("Test Finished: Server[" + speedTestResult.server.Domain + "] -> " + speedTestResult.toString())
    }

    override fun onPingStarted() {
        binding.testProgressBar.setProgress(20, true)
        binding.tvTestStatusText.text = getString(R.string.common_ping)
        binding.ivTestStatusIcon.setImageResource(R.drawable.ic_ping)
        binding.tvProgress.text = getString(R.string.common_progress)
        log("Ping Started")
    }

    override fun onPingFinished(ping: Int, jitter: Int) {
        binding.testProgressBar.setProgress(40, true)
        binding.tvTestStatusText.text = getString(R.string.test_ping_finished)
        binding.tvCurrentTestValue.text = "$ping ms | jitter: $jitter"
        log("Ping Finished: $ping ms| jitter: $jitter")
    }

    override fun onDownloadTestStarted() {
        binding.tvTestStatusText.text = getString(R.string.test_download_started)
        binding.ivTestStatusIcon.setImageResource(R.drawable.ic_download)
        log("Download Test Started")
    }

    override fun onDownloadTestProgress(progress: Int, speedMbs: Double, transferredMb: Double) {
        binding.testProgressBar.setProgress(40 + progress, true)
        binding.tvTestStatusText.text = getString(R.string.common_download)
        binding.ivTestStatusIcon.setImageResource(R.drawable.ic_download)
        binding.tvProgress.text = "$progress%"
        binding.tvCurrentTestValue.text = "$speedMbs Mb/s"
        binding.tvTrafficTestValue.text = "TransferredMb: $transferredMb"
        log("Download Test Progress: $progress% -> $speedMbs Mb/s\nTransferredMb: $transferredMb")
    }

    override fun onDownloadTestFinished(v: Double) {
        binding.tvTestStatusText.text = getString(R.string.test_download_finished)
        binding.tvCurrentTestValue.text = "$v Mb/s"
        log("Download Test Finished: $v Mb/s")
    }

    override fun onUploadTestStarted() {
        binding.tvTestStatusText.text = getString(R.string.common_upload)
        binding.ivTestStatusIcon.setImageResource(R.drawable.ic_upload)
        binding.tvProgress.text = ""
        binding.tvCurrentTestValue.text = ""
        binding.tvTrafficTestValue.text = ""
        log("Upload Test Started")
    }

    override fun onUploadTestProgress(progress: Int, speedMbs: Double, transferredMb: Double) {
        binding.testProgressBar.setProgress(140 + progress, true)
        binding.tvProgress.text = "$progress%"
        binding.tvCurrentTestValue.text = "$speedMbs Mb/s"
        binding.tvTrafficTestValue.text = "TransferredMb: $transferredMb"
        log("Upload Test Progress: $progress% -> $speedMbs Mb/s\nTransferredMb: $transferredMb")
    }

    override fun onUploadTestFinished(v: Double) {
        binding.tvCurrentTestValue.text = "$v Mb/s"
        log("Upload Test Finished: $v Mb/s")
    }

    override fun onTestWarning(s: String?) {
        binding.tvCurrentTestValue.text = getString(R.string.test_warning)
        binding.tvTrafficTestValue.text = s
        log("Test Warning: $s")
    }

    override fun onTestFatalError(s: String?) {
        binding.tvCurrentTestValue.text = getString(R.string.test_error)
        binding.tvTrafficTestValue.text = s
        binding.btnStart.isEnabled = true
        log("Test Fatal Error: $s")
    }

    override fun onTestInterrupted(s: String?) {
        binding.tvTestStatusText.text = getString(R.string.test_interrupted)
        binding.tvCurrentTestValue.text = s
        binding.btnStart.isEnabled = true
    }

    private fun log(message: String) {
        binding.tvLog.text = ("[" + sdf.format(Date(System.currentTimeMillis())) + "]: "
                + message + "\n" + binding.tvLog.text)
    }

}